lib/Sema/CSApply.cpp

```diff
@@ -2506,15 +2506,18 @@ namespace {
             auto baseTyUnwrappedName = baseTyUnwrapped->getString();
             auto loc = DSCE->getLoc();
             auto startLoc = DSCE->getStartLoc();
-            
-            tc.diagnose(loc, swift::diag::optional_ambiguous_case_ref,
-                        baseTyName, baseTyUnwrappedName, memberName.str());
-            
-            tc.diagnose(loc, swift::diag::optional_fixit_ambiguous_case_ref)
-              .fixItInsert(startLoc, "Optional");
-            tc.diagnose(loc, swift::diag::type_fixit_optional_ambiguous_case_ref,
-                        baseTyUnwrappedName, memberName.str())
-              .fixItInsert(startLoc, baseTyUnwrappedName);
+            {
+              CompoundDiagnosticTransaction transaction(tc.Diags);
+              tc.diagnose(loc, swift::diag::optional_ambiguous_case_ref,
+                          baseTyName, baseTyUnwrappedName, memberName.str());
```

Comment:
I agree, this is a really great idea! It strikes a nice balance between the two approaches I've tried so far.

Reply:
It turns out this doesn't work out too well because the transaction needs to outlive the parent InFlightDiagnostic(WithNotes). Otherwise we end up keeping the parent alive while diagnosing the notes which breaks a lot of important assumptions. 

The best alternative I can come up with at the moment is a `buildCompoundDiagnostic` helper which accepts a lambda with an implicit transaction (see my second commit). This doesn't reduce the code involved, but it's easier to read the code and understand what's going on.

```
tc.buildCompoundDiagnostic([&]() {
  tc.diagnose(...);
  tc.diagnose(...);
  tc.diagnose(...);
});
```

Reply:
Hm. I like the explicit stack scoping, but it still feels like the parent diagnostic should be in the compound action. What about this (on DiagnosticEngine)?

```c++
template<typename ...ArgTypes>
void diagnoseWithNotes(
    SourceLoc Loc, Diag<ArgTypes...> ID,
    typename detail::PassArgument<ArgTypes>::type... Args,
    llvm::function_ref<void(void)> builder) {
  CompoundDiagnosticTransaction transaction(*this);
  diagnose(Loc, ID, Args...);
  builder();
}
```

The one thing I don't like is that you can't add highlights or fix-its this way, but it would be possible to have another overload that took an InFlightDiagnostic as well. Though you'd have to call `flush()` explicitly there.

Reply:
Hmm, types are only deduced for variadic parameter packs when they come at the end of an argument list though, which makes this kind of unwieldy. On the other hand, moving builder to the front of the list would be pretty unintuitive.

I don't think the inability to add fix-its is that big an issue. If an error has notes, it very rarely has fix-its of its own.

Reply:
I ended up updating this to just accept an InFlightDiagnostic, manually specifying template parameters became too annoying.

That allows us to write:
```
tc.diagnoseWithNotes(
  tc.diagnose(...),
  [&]() {
    tc.diagnose(...);
    tc.diagnose(...);
});
```

It's not perfect, but I think it's still an improvement